official/utils/misc/distribution_utils.py

```diff
@@ -96,3 +96,51 @@ def per_device_batch_size(batch_size, num_gpus):
           ).format(num_gpus, batch_size, batch_size - remainder)
     raise ValueError(err)
   return int(batch_size / num_gpus)
+
+
+class SyntheticDataset(object):
+  """A dataset that generates synthetic data on each device."""
+
+  def __init__(self, dataset, split_by=1):
+    self._input_data = {}
+    # dataset.take(1) doesn't have GPU kernel.
+    with tf.device("device:CPU:0"):
+      tensor = tf.data.experimental.get_single_element(dataset.take(1))
+    flat_tensor = tf.nest.flatten(tensor)
+    variable_data = []
+    self._initializers = []
+    for t in flat_tensor:
+      rebatched_t = tf.split(t, num_or_size_splits=split_by, axis=0)[0]
+      assert rebatched_t.shape.is_fully_defined(), rebatched_t.shape
+      # v should be MirroredVariable.
```

Comment:
maybe remove this comment. this wont be mirrored for one device strategy, for e.g. right?